File: knowledge-server/internal/handlers/insights.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/clab/knowledge-server/internal/services"
	"github.com/clab/knowledge-server/internal/store"
	"github.com/clab/knowledge-server/internal/types"
)

// InsightsHandler handles insight extraction from task results.
type InsightsHandler struct {
	store store.KnowledgeStore
}

// NewInsightsHandler creates a new InsightsHandler.
func NewInsightsHandler(s store.KnowledgeStore) *InsightsHandler {
	return &InsightsHandler{store: s}
}

// extractRequest is the JSON body for POST /v1/insights/extract.
type extractRequest struct {
	TaskRunID string           `json:"taskRunId"`
	Result    types.TaskResult `json:"result"`
	Context   string           `json:"context"`
}

// insightOutput represents an extracted insight in the response.
type insightOutput struct {
	Type       string   `json:"type"`
	Topic      string   `json:"topic"`
	Content    string   `json:"content"`
	Tags       []string `json:"tags"`
	Confidence float64  `json:"confidence"`
}

// Decision indicator words
var decisionIndicators = []string{
	"decided", "chose", "selected", "opted",
	"switched", "migrated", "replaced", "adopted",
}

// Extract handles POST /v1/insights/extract.
// Logic: extract patterns/decisions/risks -> store as knowledge entries -> return insights.
func (h *InsightsHandler) Extract(w http.ResponseWriter, r *http.Request) {
	var req extractRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body: "+err.Error())
		return
	}

	if req.TaskRunID == "" {
		writeError(w, http.StatusBadRequest, "taskRunId is required")
		return
	}

	var insights []insightOutput
	combinedText := req.Result.Summary + " " + req.Context

	// Pattern extraction: extract keywords (top 10), if >= 3 -> create pattern insight
	keywords := services.ExtractKeywords(combinedText, 10)
	if len(keywords) >= 3 {
		insight := insightOutput{
			Type:       "pattern",
			Topic:      fmt.Sprintf("pattern:%s", req.TaskRunID),
			Content:    fmt.Sprintf("Detected pattern with keywords: %s", strings.Join(keywords, ", ")),
			Tags:       append([]string{"pattern", "extracted"}, keywords...),
			Confidence: 0.8,
		}
		insights = append(insights, insight)
	}

	// Decision extraction: check for indicator words
	lowerText := strings.ToLower(combinedText)
	for _, indicator := range decisionIndicators {
		if strings.Contains(lowerText, indicator) {
			insight := insightOutput{
				Type:       "decision",
				Topic:      fmt.Sprintf("decision:%s", req.TaskRunID),
				Content:    fmt.Sprintf("Decision detected (indicator: '%s') in task output: %s", indicator, truncate(combinedText, 200)),
				Tags:       []string{"decision", "extracted", indicator},
				Confidence: 0.8,
			}
			insights = append(insights, insight)
			break // only one decision insight per extraction
		}
	}

	// Risk extraction: if result.risks is non-empty -> create risk insight
	if len(req.Result.Risks) > 0 {
		insight := insightOutput{
			Type:       "risk",
			Topic:      fmt.Sprintf("risk:%s", req.TaskRunID),
			Content:    fmt.Sprintf("Risks identified: %s", strings.Join(req.Result.Risks, "; ")),
			Tags:       append([]string{"risk", "extracted"}, req.Result.Risks...),
			Confidence: 0.8,
		}
		insights = append(insights, insight)
	}

	// Store each insight as EXTRACTED knowledge entry with confidence 0.8
	now := time.Now().UTC().Format(time.RFC3339)
	storedCount := 0
	for _, insight := range insights {
		entry := types.KnowledgeEntry{
			ID:         newUUID(),
			Topic:      insight.Topic,
			Content:    insight.Content,
			Tags:       insight.Tags,
			Source:     "EXTRACTED",
			Confidence: insight.Confidence,
			CreatedAt:  now,
			UpdatedAt:  now,
		}
		_, err := h.store.Store(entry)
		if err == nil {
			storedCount++
		}
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"ok":        true,
		"insights":  insights,
		"stored":    storedCount,
		"taskRunId": req.TaskRunID,
	})
}

// truncate shortens a string to maxLen characters.
func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}

File: knowledge/langgraph/knowledge/insights.py
"""Insight extraction from task results -- ported from clab-platform insights.ts."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from uuid import uuid4

from langgraph.knowledge.services.keyword_extractor import extract_keywords
from langgraph.knowledge.store import KnowledgeStore
from langgraph.knowledge.types import KnowledgeEntry


@dataclass
class TaskResult:
    """Describes the outcome of a task run."""

    status: str
    summary: str
    changed_files: list[str] = field(default_factory=list)
    risks: list[str] = field(default_factory=list)
    followups: list[str] = field(default_factory=list)


@dataclass
class ExtractedInsight:
    """A single insight extracted from a task run."""

    id: str
    task_run_id: str
    type: str  # "pattern" | "decision" | "risk" | "learning"
    title: str
    description: str
    evidence: list[str]
    tags: list[str]
    created_at: str


DECISION_INDICATORS: list[str] = [
    "decided",
    "chose",
    "selected",
    "opted",
    "switched",
    "migrated",
    "replaced",
    "adopted",
]


async def extract_insights(
    task_run_id: str,
    result: TaskResult,
    context: str,
    store: KnowledgeStore,
) -> list[ExtractedInsight]:
    """Extract insights from a task run result and store them.

    1. Pattern detection: extract keywords, if >= 3 keywords -> create pattern insight
    2. Decision detection: check summary for decision indicator words -> create decision insight
    3. Risk detection: if result.risks is non-empty -> create risk insight
    4. Store each insight as a knowledge entry (source=EXTRACTED, confidence=0.8)
    5. Return list of extracted insights
    """

    extracted: list[ExtractedInsight] = []
    combined_text = f"{result.summary} {context}"

    # 1. Pattern detection
    keywords = extract_keywords(combined_text, max_count=6)

    if len(keywords) >= 3:
        now = datetime.now(tz=timezone.utc).isoformat()
        extracted.append(
            ExtractedInsight(
                id=str(uuid4()),
                task_run_id=task_run_id,
                type="pattern",
                title=f"Pattern: {', '.join(keywords[:3])}",
                description=f"Recurring themes detected in task output: {', '.join(keywords)}",
                evidence=[result.summary[:500]],
                tags=keywords,
                created_at=now,
            )
        )

    # 2. Decision detection
    summary_lower = result.summary.lower()
    has_decision = any(ind in summary_lower for ind in DECISION_INDICATORS)

    if has_decision:
        now = datetime.now(tz=timezone.utc).isoformat()
        extracted.append(
            ExtractedInsight(
                id=str(uuid4()),
                task_run_id=task_run_id,
                type="decision",
                title=f"Decision recorded from task {task_run_id}",
                description=result.summary[:500],
                evidence=[result.summary[:500]],
                tags=["decision", *keywords[:3]],
                created_at=now,
            )
        )

    # 3. Risk detection
    if result.risks:
        now = datetime.now(tz=timezone.utc).isoformat()
        extracted.append(
            ExtractedInsight(
                id=str(uuid4()),
                task_run_id=task_run_id,
                type="risk",
                title=f"{len(result.risks)} risk(s) identified",
                description="; ".join(result.risks),
                evidence=list(result.risks),
                tags=["risk", *keywords[:2]],
                created_at=now,
            )
        )

    # 4. Store each insight as a knowledge entry
    for insight in extracted:
        now = datetime.now(tz=timezone.utc).isoformat()
        entry = KnowledgeEntry(
            id=str(uuid4()),
            topic=insight.title,
            content=insight.description,
            tags=insight.tags,
            source="EXTRACTED",
            confidence=0.8,
            created_at=now,
        )
        await store.store(entry)

    return extracted
